---
title: Rey et al. 2020 data analysis
date: "`r Sys.Date()`"
author: Pieter Provoost
output: (function(...) {
  rmdformats::robobook(toc_depth = 4, pandoc_args = c("+RTS", "-K2000m", "-RTS"), ...) })
editor_options: 
  chunk_output_type: console
knit: (function(inputFile, encoding) {
  rmarkdown::render(inputFile, encoding = encoding, output_file = "../../../docs/rey_analysis.html") })  
---

<!--
- https://deneflab.github.io/MicrobeMiseq/demos/mothur_2_phyloseq.html
- https://joey711.github.io/phyloseq/plot_ordination-examples.html
- https://david-barnett.github.io/microViz/articles/web-only/ordination.html
- https://www.gdc-docs.ethz.ch/MDA/handouts/MDA20_PhyloseqFormation_Mahendra_Mariadassou.pdf
- http://rstudio-pubs-static.s3.amazonaws.com/266780_cac4994322494658904507a7606b1dd8.html
-->

<style>
img { box-shadow: none !important; border: 0px !important; }
</style>

In this notebook we will analyze PacMAN pipeline results for a subset of COI samples from Rey, A., Basurko, O. C., &#38; Rodriguez-Ezpeleta, N. (2020). Considerations for metabarcoding-based port biological baseline surveys aimed at marine nonindigenous species monitoring and risk assessments. Ecology and Evolution, 10(5), 2452â€“2465. https://doi.org/10.1002/ece3.6071

## Running this notebook

To run this notebook in its entirety or to have access to the data files used in this notebook, clone or download and extract the [iobis/pacman-pipeline-training repository](https://github.com/iobis/pacman-pipeline-training). The notebook can be found in the `datasets/rey/analysis` directory, open this directory in RStudio after downloading (`File` > `Open Project` or navigate to Desktop in the Files panel and select `More` > `Set As Working Directory`).

If you have trouble locating the Desktop folder in RStudio, set your Desktop as the default working directory using `Tools` > `Global Options` > `Default working directory`.

## Importing data

The PacMAN pipeline exports results as a `phyloseq` object which can be imported into R for analysis. First install the phyloseq package from Bioconductor:

```{r message=FALSE, warning=FALSE, eval=FALSE}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("phyloseq")
```

This is how `phyloseq` objects are structured:

![Figure from <https://joey711.github.io/phyloseq/import-data.html>](phyloseq.png)

Read the `phyloseq` object and verify that we have:

- a OTU table
- a sample table
- a taxonomy table

A phylogenetic tree has not been calculated so that slot is empty.

```{r message=FALSE, warning=FALSE}
physeq <- readRDS("../results_noblast/phyloseq_object.rds")

physeq
```

While we can access these tables individually (for example using `physeq@sam_data` or `sample_data(physeq)`), the phyloseq package also has a function `psmelt()` to convert the `phyloseq` object into a single large data frame:

```{r message=FALSE, warning=FALSE}
library(dplyr)
library(phyloseq)

df <- psmelt(physeq) %>%
  # convert empty strings to NA across all character columns
  mutate_if(is.character, ~na_if(., "")) %>%
  # convert to tibble for prettier printing
  as_tibble()

df
```

## Exploratory data analysis

### Abundance by phylum

Let's first determine which are the most abundant phyla across all samples. We will use this information to bin the less abundant phyla into a category `Other` for simplifying some of our graphics.

```{r message=FALSE, warning=FALSE}
stats_phyla <- df %>%
  group_by(phylum) %>%
  summarize(abundance = sum(Abundance)) %>%
  arrange(desc(abundance))

top_phyla <- head(na.omit(stats_phyla$phylum), 8)
top_phyla
```

Now create a color scale for these phyla using the `RColorBrewer` package:

```{r message=FALSE, warning=FALSE}
phylum_colors <- RColorBrewer::brewer.pal(9, "Spectral")
names(phylum_colors) <- c(top_phyla, "Other")
phylum_colors
```

And add a new column with binned phyla:

```{r message=FALSE, warning=FALSE}
df <- df %>%
  mutate(phylum_binned = ifelse(is.na(phylum) | phylum %in% top_phyla, phylum, "Other")) %>%
  mutate(phylum_binned = factor(phylum_binned, levels = c(top_phyla, "Other")))
```

Let's also add the binned phyla to the `phyloseq` object for later use:

```{r message=FALSE, warning=FALSE}
# remotes::install_github("david-barnett/microViz")

physeq <- physeq %>%
  microViz::tax_mutate(phylum = ifelse(phylum == "", NA, phylum)) %>%
  microViz::tax_mutate(phylum_binned = ifelse(is.na(phylum) | phylum %in% top_phyla, phylum, "Other"))
```

### Abundance by sample type and phylum

First list the abundance by phylum and sample type:

```{r message=FALSE, warning=FALSE}
library(tidyr)

stats_type_phylum <- df %>%
  group_by(eventRemarks, phylum) %>%
  summarize(abundance = sum(Abundance))

spread(stats_type_phylum, eventRemarks, abundance) %>%
  mutate(total = `settlement plates` + `filtered water`) %>%
  arrange(desc(total)) %>%
  knitr::kable()
```

Now create a graph using the binned phyla:

```{r message=FALSE, warning=FALSE}
library(ggplot2)

stats_type_phylum <- df %>%
  # calculate total abundance by sample type
  group_by(eventRemarks) %>%
  mutate(abundance = Abundance / sum(Abundance)) %>%
  group_by(eventRemarks, phylum_binned) %>%
  summarize(abundance = sum(abundance))
  
ggplot() +
  geom_bar(data = stats_type_phylum, aes(y = eventRemarks, fill = phylum_binned, x = abundance), stat = "identity") +
  scale_fill_manual(values = phylum_colors, na.value = "#eeeeee") +
  theme_minimal()
```

### Abundance (reads) by sample and phylum

```{r message=FALSE, warning=FALSE}
stats_sample_phylum <- df %>%
  # calculate total abundance by sample type
  group_by(Sample) %>%
  mutate(relative_abundance = Abundance / sum(Abundance)) %>%
  group_by(Sample, eventRemarks, phylum_binned) %>%
  summarize(relative_abundance = sum(relative_abundance), abundance = sum(Abundance))
```

Absolute reads abundance:

```{r message=FALSE, warning=FALSE}
ggplot() +
  geom_bar(data = stats_sample_phylum, aes(y = Sample, fill = phylum_binned, x = abundance), stat = "identity") +
  scale_fill_manual(values = phylum_colors, na.value = "#eeeeee") +
  theme_minimal() +
  theme(panel.grid.major.y = element_blank(), panel.grid.minor.y = element_blank())
```

Relative reads abundance:

```{r message=FALSE, warning=FALSE}
ggplot() +
  geom_bar(data = stats_sample_phylum, aes(y = Sample, fill = phylum_binned, x = relative_abundance), stat = "identity") +
  scale_fill_manual(values = phylum_colors, na.value = "#eeeeee") +
  theme_minimal() +
  theme(panel.grid.major.y = element_blank(), panel.grid.minor.y = element_blank())
```

### Most abundant species

Let's list the species with the highest relative abundance across all samples:

```{r message=FALSE, warning=FALSE}
top_species <- df %>%
  filter(!is.na(species)) %>%
  group_by(species) %>%
  summarize(abundance = sum(Abundance)) %>%
  arrange(desc(abundance)) %>%
  head(20)

top_species %>% knitr::kable()
```

For the most abundant species, plot the abundance by sample:

```{r message=FALSE, warning=FALSE}
stats <- df %>%
  filter(species %in% top_species$species) %>%
  group_by(species, eventRemarks, Sample) %>%
  summarize(abundance = sum(Abundance))

ggplot() +
  geom_jitter(data = stats, aes(y = species, x = abundance, color = eventRemarks), pch = 21, height = 0.1, width = 0) +
  scale_color_brewer(palette = "Set1") +
  theme_minimal()
```

## Invasive species

In this section we will check our results against the World Register of Introduced Marine Species (WRiMS). The repository contains a CSV file containing the LSIDs for all species in WRiMS. Read this CSV and use the LSIDs to only retain species from WRiMS.

```{r message=FALSE, warning=FALSE}
wrims_lsids <- read.csv("wrims_lsids.csv")

invasives <- df %>%
  filter(lsid %in% wrims_lsids$lsid)

invasives %>%
  group_by(phylum, species) %>%
  summarize(abundance = sum(Abundance)) %>%
  arrange(desc(abundance)) %>%
  rmarkdown::paged_table()
```

## Alpha diversity

Use `vegan::plot_richness()` to visualize the number of ASVs per sample:

```{r message=FALSE, warning=FALSE}
plot_richness(physeq, measures = c("Observed"), color = "eventRemarks")
```

The vegan package can also calculate a number of biodiversity indices. But keep in mind that some of these will use the reads abundance which does not necessarily correlate with actual abundance in terms of individuals, such as the Chao1 estimator which uses the numbers of singletons and doubletons to obtain a lower bound for the expected asymptotic species richness, or Shannon entropy which uses relative abundance.

```{r message=FALSE, warning=FALSE}
estimate_richness(physeq)
```

## ASV accumulation curves

Rarefaction is an interpolation of species richness to smaller samples sizes to allow comparisons between samples.

```{r message=FALSE, warning=FALSE}
vegan::rarecurve(t(otu_table(physeq)), step = 1000)
```

Alternatively, run `vegan::rarefy()` for every sample and construct the graph yourself:

```{r message=FALSE, warning=FALSE}
community <- t(otu_table(physeq))

curves <- purrr::map(physeq@sam_data$eventID, function(sample) {
  community_row <- community[sample,]
  sample_sizes <- seq(1000, sum(community_row), by = 1000)
  rarefied <- vegan::rarefy(community_row, sample_sizes)
  curve <- data.frame(
    eventID = sample,
    sample_size = sample_sizes,
    asvs = as.numeric(rarefied)
  ) %>%
    left_join(physeq@sam_data, by = "eventID")
}) %>%
  bind_rows()

samples <- curves %>%
  group_by(eventID) %>%
  summarize(sample_size = max(sample_size), asvs = max(asvs))

ggplot() +
  geom_line(data = curves %>% group_by(eventID), aes(x = sample_size, y = asvs, color = eventRemarks, group = eventID)) +
  geom_label(data = samples, aes(sample_size, asvs, label = eventID), size = 2) +
  scale_color_brewer(palette = "Set1") +
  theme_classic()
```

## Beta diversity

### Multidimensional scaling with phyloseq

Phyloseq's `ordinate()` function allows us to do MultiDimensional Scaling (MDS) or Principal Coordinate Analysis (PCoA) on our dataset. The default for this function is to use [Bray-Curtis Dissimilarity](https://en.wikipedia.org/wiki/Bray%E2%80%93Curtis_dissimilarity) to calculate a distance matrix.

```{r message=FALSE, warning=FALSE}
ord <- ordinate(physeq = physeq, method = "PCoA", distance = "bray")

plot_ordination(physeq = physeq, ordination = ord, type = "samples", color = "eventRemarks", shape = "locality") +
  geom_point(aes(color = eventRemarks), alpha = 1, size = 4) +
  geom_text(aes(label = materialSampleID), alpha = 0.7, size = 3, vjust = 2) +
  stat_ellipse(aes(group = eventRemarks)) +
  scale_color_brewer(palette = "Set1") +
  scale_shape(solid = TRUE) +
  theme_classic()
```

We can also add taxa to the ordination plot, but styling this is a bit awkward as samples and taxa are treated as a single dataset.

```{r message=FALSE, warning=FALSE}
plot_ordination(physeq = subset_taxa(physeq, !is.na(phylum)), ordination = ord, type = "biplot", color = "phylum_binned") +
  scale_color_manual(values = phylum_colors, na.value = "#000000") +
  geom_text(aes(label = materialSampleID), alpha = 0.7, size = 3, vjust = 2) +
  theme_classic()
```

### Non-metric MDS (NMDS) with phyloseq

Another ordination method is non-metric MDS (NMDS). Let's try NMDS with our dataset aggregated at the species level, and plot the most abundant species together with the sites.

First aggregates the `phyloseq` object to the species level:

```{r message=FALSE, warning=FALSE}
physeq_species <- tax_glom(physeq, taxrank = "species", NArm = TRUE, bad_empty = c(NA, ""))
```

Then perform the ordination with method `NMDS`:

```{r message=FALSE, warning=FALSE, results="hide"}
ord_species <- ordinate(physeq_species, method = "NMDS", distance = "bray")
```

Extract the NMDS values for the sites and species, and join with the relevant tables from the `phyloseq` object to get the necessary metadata for plotting (such as species name, locality, and sample type):

```{r message=FALSE, warning=FALSE}
sites <- ord_species$points %>%
  as.data.frame() %>%
  tibble::rownames_to_column("site") %>%
  inner_join(physeq_species@sam_data %>% as.data.frame() %>% tibble::rownames_to_column("site"), by = "site")

species <- ord_species$species %>%
  as.data.frame() %>%
  tibble::rownames_to_column("asv") %>%
  left_join(physeq_species@tax_table %>% as.data.frame() %>% tibble::rownames_to_column("asv"), by = "asv") %>%
  filter(species %in% top_species$species)
```

```{r message=FALSE, warning=FALSE}
ggplot() +
  geom_segment(data = species, aes(x = 0, y = 0, xend = MDS1, yend = MDS2), arrow = arrow(length = unit(2, "mm")), color = "#aaaaaa") +
  geom_point(data = sites, aes(MDS1, MDS2, color = eventRemarks, shape = locality), size = 3, stroke = 1.5) +
  geom_text(data = species, aes(MDS1, MDS2, label = species), size = 3, vjust = "outward") +
  scale_color_brewer(palette = "Set1") +
  scale_shape(solid = FALSE) +
  theme_classic()
```
